config/configoptional/optional.go

```diff
@@ -0,0 +1,119 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package configoptional // import "go.opentelemetry.io/collector/config/configoptional"
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.opentelemetry.io/collector/confmap"
+)
+
+// Optional represents a value that may or may not be present.
+// It supports three flavors: Some(value), Default(value), and None.
+// The zero value of Optional is None.
+type Optional[T any] struct {
+	// Enabled indicates if the value is present.
+	Enabled bool `mapstructure:"enabled"`
```

Comment:
Left a response on the original PR https://github.com/open-telemetry/opentelemetry-collector/pull/13044#issuecomment-2898090891. IMO adding `enabled` directly to the optional type is an abstraction leak, it does not need to be done at the framework level, the user always has the option of adding it explicitly to `T`.